cmd/geth/main.go

```diff
@@ -286,6 +286,20 @@ func prepare(ctx *cli.Context) {
 	case !ctx.GlobalIsSet(utils.NetworkIdFlag.Name):
 		log.Info("Starting Geth on Ethereum mainnet...")
 	}
+
+	// Load the config file early to check the Cache configuration value
+	cfg := gethConfig{}
+	if file := ctx.GlobalString(configFileFlag.Name); file != "" {
+		if err := loadConfig(file, &cfg); err != nil {
+			utils.Fatalf("%v", err)
+		}
+	}
```

Comment:
I'm not sure why the config needs to be loaded another time here. Maybe it's because the code below, which post-processes the cache amount automatically is in the wrong place. 

There are two parts to this post-processing. First we have some logic to set the default value based on blockchain if it is not specified as a flag. This part is only related to flags and doesn't need the config file value. Then there is the part where the cache amount is 'sanitized'. We need to apply that to the config file as well.
Maybe it should be moved to cmd/utils/flags.go.

Reply:
The config file is loaded first here since this code is executed before the config file is normally processed.

From my understanding the problem is the cache has to be set early here before the p2p networking is setup.

I'm still new to the codebase so please let me know how you'd prefer it to be implemented.

Reply:
Geth startup works as follows (see it [here](https://github.com/ethereum/go-ethereum/blob/91726e8aadeebc4bda8f136aebf4b5c7da8b9d74/cmd/geth/config.go#L110))*.

1. pre-setup (everything that runs before `makeConfigNode`)
2. config file is loaded
3. flags related to `node.Config` are processed and override values in config
4. `node.Node` object is constructed
5. flags related to `ethconfig.Config` are processed and override values in config
6. `eth.Ethereum` object is constructed and registered in node
7. node is started

I think this code here runs during the pre-setup, which is the wrong place to adjust anything in the config.
To make it work properly, this code needs to run during phase 3 or phase 5. The code for those phases is in package cmd/utils. This is why I suggest that the adjusting of the cache amount should be moved there.

Reply:
When we discussed this PR in the team today, someone said that adding the `Cache` field to `node.Config` is a bad idea. I agree, because we already have config fields for the cache amount in `ethconfig.Config`: https://github.com/ethereum/go-ethereum/blob/91726e8aadeebc4bda8f136aebf4b5c7da8b9d74/eth/ethconfig/config.go#L72

So what really needs to happen here is: the sanity checking of the database cache amount needs to be moved into phase 5, i.e. [in the `SetEthConfig` function](https://github.com/ethereum/go-ethereum/blob/91726e8aadeebc4bda8f136aebf4b5c7da8b9d74/cmd/utils/flags.go#L1447)